File: matlab/generate_database.m
clear all;

%% Initialize params

init_params;

%% Simulation parameters
step_time = 0.00001; % Simulation step time
sim_time = 10;   % Simulation time in seconds

%% Generate time vector
time = 0:step_time:sim_time;

%% Tamanho do dataset de treino
n_inputs = 2;
n_outputs = 2;
N = length(time);
outputs = zeros(N, n_outputs);
inputs = zeros(N, n_inputs);

%% Open Simulink model
model_name = 'gmaw_process';
open_system(model_name);

%% Setup Signal Builder block
f_lv = 0*fo; % Lower value for the binary signal
f_uv = fo; % Upper value for the binary signal
Ir_lv = 0.6*Ir; % Lower value for the binary signal
Ir_uv = Ir; % Upper value for the binary signal

%% Generate a pseudo-random binary signal using the parameters
f_binary_signal = randi([0, 1], 1, N);
Ir_binary_signal = randi([0, 1], 1, N);

f_signal = f_lv + (f_uv - f_lv) * f_binary_signal;
Ir_signal = Ir_lv + (Ir_uv - Ir_lv) * Ir_binary_signal;

inputs(:, 1) = f_signal';
inputs(:,2) = Ir_signal';

%% Setup Signal Builder block in the model
signal_f_path = [model_name, '/Signal f']; % Update with your Signal Builder block's path
signal_f_block = get_param(signal_f_path, 'Handle');
% Set the generated binary signal using the 'set' function
signalbuilder(signal_f_block, 'set', 1, 1, time, f_signal);

signal_Ir_path = [model_name, '/Signal Ir']; % Update with your Signal Builder block's path
signal_Ir_block = get_param(signal_Ir_path, 'Handle');
% Set the generated binary signal using the 'set' function
signalbuilder(signal_Ir_block, 'set', 1, 1, time, Ir_signal);

%% Simulate
% Define simulation options structure
simOut = sim(model_name, 'FixedStep', num2str(step_time), 'StopTime', num2str(sim_time));

tout = simOut.tout;
we = simOut.we.Data;
we = resample(we', tout, 1/step_time);
h = simOut.h.Data;
h = resample(h', tout, 1/step_time);

% Generate Gaussian noise
noise_mean = 0;      % Mean of the Gaussian noise
noise_stddev = 1e-6;  % Standard deviation of the Gaussian noise
gaussian_noise = noise_mean + noise_stddev * randn(size(time));

we = we + gaussian_noise;
h = h + gaussian_noise;

outputs(:,1) = we;
outputs(:,2) = h;

%% Plot
% figure;
% subplot(2,1,1);
% plot(time, f_signal);
% xlabel('t');
% ylabel('Ir');
% title('Input f');
% subplot(2,1,2);
% plot(time, Ir_signal);
% xlabel('t');
% ylabel('Ir');
% title('Input Ir');
% sgtitle('Inputs', 'FontSize', 16); % Adjust font size if needed
% 
% figure;
% subplot(2,1,1);
% plot(time, we);
% xlabel('t');
% ylabel('we');
% title('Output we');
% subplot(2,1,2);
% plot(time, h);
% xlabel('t');
% ylabel('h');
% title('Output h');
% sgtitle('Outputs', 'FontSize', 16); % Adjust font size if needed

%% Save data
% Define column headers
inputs_headers = {'f', 'Ir'};
outputs_headers = {'we', 'h'};

% Convert data array to a table
inputs_table = array2table(inputs, 'VariableNames', inputs_headers);
outputs_table = array2table(outputs, 'VariableNames', outputs_headers);

% Save data table to a CSV file
writetable(inputs_table, 'C:/Users/lucas/OneDrive/Documentos/GitHub/lstm-control-waam/database/train_inputs.csv');
writetable(outputs_table, 'C:/Users/lucas/OneDrive/Documentos/GitHub/lstm-control-waam/database/train_outputs.csv');

%% Generate a pseudo-random binary signal using the parameters

n_amplitudes = 10;
f_binary_signal = randi([0, n_amplitudes-1], 1, N)/(n_amplitudes-1);
Ir_binary_signal = randi([0, n_amplitudes-1], 1, N)/(n_amplitudes-1);

f_signal = f_lv + (f_uv - f_lv) * f_binary_signal;
Ir_signal = Ir_lv + (Ir_uv - Ir_lv) * Ir_binary_signal;

inputs(:, 1) = f_signal';
inputs(:,2) = Ir_signal';

%% Setup Signal Builder block in the model
signal_f_path = [model_name, '/Signal f']; % Update with your Signal Builder block's path
signal_f_block = get_param(signal_f_path, 'Handle');
% Set the generated binary signal using the 'set' function
signalbuilder(signal_f_block, 'set', 1, 1, time, f_signal);

signal_Ir_path = [model_name, '/Signal Ir']; % Update with your Signal Builder block's path
signal_Ir_block = get_param(signal_Ir_path, 'Handle');
% Set the generated binary signal using the 'set' function
signalbuilder(signal_Ir_block, 'set', 1, 1, time, Ir_signal);

%% Simulate
% Define simulation options structure
simOut = sim(model_name, 'FixedStep', num2str(step_time), 'StopTime', num2str(sim_time));

tout = simOut.tout;
we = simOut.we.Data;
we = resample(we', tout, 1/step_time);
h = simOut.h.Data;
h = resample(h', tout, 1/step_time);

% Generate Gaussian noise
noise_mean = 0;      % Mean of the Gaussian noise
noise_stddev = 1e-6;  % Standard deviation of the Gaussian noise
gaussian_noise = noise_mean + noise_stddev * randn(size(time));

we = we + gaussian_noise;
h = h + gaussian_noise;

outputs(:,1) = we;
outputs(:,2) = h;

%% Plot
% figure;
% subplot(2,1,1);
% plot(time, f_signal);
% xlabel('t');
% ylabel('Ir');
% title('Input f');
% subplot(2,1,2);
% plot(time, Ir_signal);
% xlabel('t');
% ylabel('Ir');
% title('Input Ir');
% sgtitle('Inputs', 'FontSize', 16); % Adjust font size if needed
% 
% figure;
% subplot(2,1,1);
% plot(time, we);
% xlabel('t');
% ylabel('we');
% title('Output we');
% subplot(2,1,2);
% plot(time, h);
% xlabel('t');
% ylabel('h');
% title('Output h');
% sgtitle('Outputs', 'FontSize', 16); % Adjust font size if needed

%% Save data
% Define column headers
inputs_headers = {'f', 'Ir'};
outputs_headers = {'we', 'h'};

% Convert data array to a table
inputs_table = array2table(inputs, 'VariableNames', inputs_headers);
outputs_table = array2table(outputs, 'VariableNames', outputs_headers);

% Save data table to a CSV file
writetable(inputs_table, 'C:/Users/lucas/OneDrive/Documentos/GitHub/lstm-control-waam/database/test_inputs.csv');
writetable(outputs_table, 'C:/Users/lucas/OneDrive/Documentos/GitHub/lstm-control-waam/database/test_outputs.csv');

File: python/generate_database.py
from python.gmaw_process import *
import numpy as np
import pandas as pd

# Paths
data_dir = "database/"

# Simulation parameters
step_time = 1e-05  # Simulation step time
sim_time = 1e-02  # Simulation time in seconds

# Generate time vectors
time = np.arange(0, sim_time + step_time, step_time).round(
    int(np.log10(1 / step_time))
)

# Dataset size
n_inputs = 2
n_outputs = 2
N = len(time)

# Input bounds
f_lv = 0  # Lower value for the binary signal
f_uv = fo  # Upper value for the binary signal
Ir_lv = 0.6 * Ir  # Lower value for the binary signal
Ir_uv = Ir  # Upper value for the binary signal
input_bounds = (f_lv, f_uv, Ir_lv, Ir_uv)


def generate_dataset(N, input_bounds, n_classes, x0):
    inputs = np.zeros((N, n_inputs))
    outputs = np.zeros((N, n_outputs))

    f_lv, f_uv, Ir_lv, Ir_uv = input_bounds
    # Generate a pseudo-random binary signal using the parameters
    f_binary_signal = np.random.randint(0, n_classes, size=N)
    Ir_binary_signal = np.random.randint(0, n_classes, size=N)
    f_signal = f_lv + (f_uv - f_lv) * f_binary_signal
    Ir_signal = Ir_lv + (Ir_uv - Ir_lv) * Ir_binary_signal
    inputs[:, 0] = f_signal
    inputs[:, 1] = Ir_signal

    outputs[0, :] = gmaw_outputs(x0)
    x = x0
    for i in range(1, time.shape[0]):
        u = inputs[i - 1, :]
        x = solve_rk4(gmaw_states, x, step_time, u)
        y = gmaw_outputs(x)
        outputs[i, :] = y
    return inputs, outputs


# initial state
x0 = [1e-10, 1e-10]
inputs_train, outputs_train = generate_dataset(N, input_bounds, 2, x0)
inputs_test, outputs_test = generate_dataset(N, input_bounds, 10, x0)

inputs_train = pd.DataFrame(inputs_train, columns=["f", "Ir"])
inputs_test = pd.DataFrame(inputs_test, columns=["f", "Ir"])
outputs_train = pd.DataFrame(outputs_train, columns=["we", "h"])
outputs_test = pd.DataFrame(outputs_test, columns=["we", "h"])

inputs_train.to_csv(data_dir + "inputs_train.csv", index=False)
inputs_test.to_csv(data_dir + "inputs_test.csv", index=False)
outputs_train.to_csv(data_dir + "outputs_train.csv", index=False)
outputs_test.to_csv(data_dir + "outputs_test.csv", index=False)
